Support lpc55s69 usb FS controller

MarkWangChinese opened the pull request:
Please reference the hal_nxp PR https://github.com/zephyrproject-rtos/hal_nxp/pull/9

jfischer-phytec-iot commented on drivers/usb/device/usb_dc_mcux_ip3511fs.c:
```diff
@@ -0,0 +1,528 @@
+	}
+
+	if (EP_ADDR2IDX(ep) == USB_ENDPOINT_CONTROL) {
+		if (s_Device.setupDataStage == SETUP_DATA_STAGE_DONE) {
```
What is setupDataStage used for?

jfischer-phytec-iot replied:
NXP controller driver has a build-in USB device stack? Please describe it widely.

MarkWangChinese replied:
Yes, NXP has USB stack in the MCUXpresso SDK package. I take the part (controller driver) and implement one shime for Zephyr.

jfischer-no reviewed (changes requested):
Please fix style issues. Please also fix style issues in [usb_dc_mcux_ehci](https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/usb/device/usb_dc_mcux_ehci.c). Follow [Zephyr Project Coding Style](https://docs.zephyrproject.org/latest/contribute/index.html#coding-style)

MaureenHelm commented on dts/arm/nxp/nxp_lpc55S6x.dtsi:
```diff
@@ -80,6 +80,16 @@
 		};
+
+		usbd:usbd@40084000 {
+			compatible = "nxp,kinetis-usbd";
```
You need a new compatible since this is different hardware from Kinetis

MarkWangChinese replied:
Thanks Maureen, I will update it. There are many failed cases in shippable because there are changes in the hal_nxp. How to sync the hal_nxp's PR to this PR and let the Shippable to get the correct codes of hal_nxp?

MaureenHelm replied:
In `west.yml`, set the hal_nxp revision to `pull/9/head`

MaureenHelm reviewed (changes requested):
You need to update west.yaml to pick up zephyrproject-rtos/hal_nxp#9

agansari commented on soc/arm/nxp_lpc/lpc55xxx/soc.c:
```diff
@@ -59,6 +59,19 @@ static ALWAYS_INLINE void clkInit(void)
+	CLOCK_AttachClk(kFRO_HF_to_USB0_CLK);
+	/* enable usb0 host clock */
+	CLOCK_EnableClock(kCLOCK_Usbhsl0);
+	/*According to reference mannual, device mode setting has to be set by access usb host register */
```
typo mannual

MarkWangChinese replied:
Fixed

agansari reviewed (approved):
Looks good, tested on my hardware.

MarkWangChinese committed:
drivers: usb: add NXP IP3511fs USB device controller driver shim

usb_dc_mcux_ip3511fs.c driver is one shim of the NXP SDK ip3511 driver

Signed-off-by: Mark Wang <yichang.wang@nxp.com>

MarkWangChinese committed:
soc: nxp_imx: configure USB device ip3511 controller driver

add usbd definition to lpc556x dts file and add new compatible file,
set IP3511 controller default value in LPC55xxx default config file,
add IP3511 controller driver MACROs to dts_fixup.h,
initialize IP3511 clock in rt soc.c
add HAS_MCUX_USB_IP3511FS for supported soc in Kconfig.soc

Signed-off-by: Mark Wang <yichang.wang@nxp.com>

MarkWangChinese commented on boards/arm/lpcxpresso55s69/doc/index.rst:
```diff
@@ -64,6 +64,8 @@ features:
 +-----------+------------+-------------------------------------+
```
add the commit for index.rst change.

dbkinder commented on boards/arm/lpcxpresso55s69/doc/index.rst:
```diff
@@ -104,6 +108,13 @@ Serial Port
+USB
+===
+
+The LPC55S69 SoC has two USB device controllers that supports
```
``supports`` on this line and the next should be ``support``

dbkinder commented on boards/arm/lpcxpresso55s69/doc/index.rst:
```diff
@@ -104,6 +108,13 @@ Serial Port
+
+The LPC55S69 SoC has two USB device controllers that supports
+device functions and two USB host controllers that supports host functions.
+Only USB device function is supported in Zephyr at the moment.
```
I'd delete "at the moment".

MarkWangChinese committed:
boards: mimxrt1050_evk: configure USB device EHCI driver for rt1050

add usb description to lpc55s69's index.rst
set usbd status as OK in board dts file.
add usb_device supported information in yaml file.

Signed-off-by: Mark Wang <yichang.wang@nxp.com>

MarkWangChinese committed:
west.yml: update nxp_hal SHA to ip3511 support PR

the hal_nxp ip3511 driver PR is #9.

Signed-off-by: Mark Wang <yichang.wang@nxp.com>

MarkWangChinese commented:
> Please fix style issues. Please also fix style issues in [usb_dc_mcux_ehci](https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/usb/device/usb_dc_mcux_ehci.c). Follow [Zephyr Project Coding Style](https://docs.zephyrproject.org/latest/contribute/index.html#coding-style)

jfischer-no reviewed (changes requested):
Please fix style issues! Please also fix style issues in [usb_dc_mcux_ehci](https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/usb/device/usb_dc_mcux_ehci.c). Follow [Zephyr Project Coding Style](https://docs.zephyrproject.org/latest/contribute/index.html#coding-style)
```

./scripts/checkpatch.pl -g HEAD~3
a6bfbe632786f7895af569540cf101f670f744e2:80: WARNING:LONG_LINE: line over 80 characters
#80: FILE: drivers/usb/device/usb_dc_mcux_ip3511fs.c:31:
+#define EP_ABS_IDX(ep)   (((ep) & ~USB_EP_DIR_MASK) * 2 + (((ep) & USB_EP_DIR_MASK) >> 7))

a6bfbe632786f7895af569540cf101f670f744e2:93: WARNING:LONG_LINE: line over 80 characters
#93: FILE: drivers/usb/device/usb_dc_mcux_ip3511fs.c:44:
+static const usb_device_controller_interface_struct_t s_UsbDevice3511IpInterface = {

a6bfbe632786f7895af569540cf101f670f744e2:94: WARNING:LONG_LINE: line over 80 characters
#94: FILE: drivers/usb/device/usb_dc_mcux_ip3511fs.c:45:
+       USB_DeviceLpc3511IpInit, USB_DeviceLpc3511IpDeinit, USB_DeviceLpc3511IpSend,

a6bfbe632786f7895af569540cf101f670f744e2:95: WARNING:LONG_LINE: line over 80 characters
#95: FILE: drivers/usb/device/usb_dc_mcux_ip3511fs.c:46:
+       USB_DeviceLpc3511IpRecv, USB_DeviceLpc3511IpCancel, USB_DeviceLpc3511IpControl

a6bfbe632786f7895af569540cf101f670f744e2:117: WARNING:LONG_LINE: line over 80 characters
#117: FILE: drivers/usb/device/usb_dc_mcux_ip3511fs.c:68:
+       s_Device.interface = (const usb_device_controller_interface_struct_t *)(&s_UsbDevice3511IpInterface);

a6bfbe632786f7895af569540cf101f670f744e2:119: WARNING:LONG_LINE: line over 80 characters
#119: FILE: drivers/usb/device/usb_dc_mcux_ip3511fs.c:70:
+                                                                 &s_Device, &s_Device.controllerHandle)) {

a6bfbe632786f7895af569540cf101f670f744e2:217: WARNING:LONG_LINE: line over 80 characters
#217: FILE: drivers/usb/device/usb_dc_mcux_ip3511fs.c:168:
+                                                 kUSB_DeviceControlEndpointDeinit,

a6bfbe632786f7895af569540cf101f670f744e2:221: WARNING:LONG_LINE: line over 80 characters
#221: FILE: drivers/usb/device/usb_dc_mcux_ip3511fs.c:172:
+                                         kUSB_DeviceControlEndpointInit, &epInit);
```
and so on

carlescufi commented:
@MarkWangChinese can you please address the style and review comments?
CC @MaureenHelm

MarkWangChinese commented:
> @MarkWangChinese can you please address the style and review comments?
> CC @MaureenHelm

Sorry for the later reply, I will work on it. Do you mean that "WARNING:LONG_LINE" must be fixed too?

jfischer-phytec-iot commented:
> > @MarkWangChinese can you please address the style and review comments?
> > CC @MaureenHelm
> 
> Sorry for the later reply, I will work on it. Do you mean that "WARNING:LONG_LINE" must be fixed too?

Please fix coding style issues. Follow [Zephyr Project Coding Style](https://docs.zephyrproject.org/latest/contribute/index.html#coding-style) Please take a look at #21175

jfischer-no reviewed (changes requested):
Please rebase and resolve conflicts.
